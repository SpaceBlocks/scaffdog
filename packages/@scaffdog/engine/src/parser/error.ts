import type { SourceRange } from '@scaffdog/types';

export type ParseErrorEntry = {
  message: string;
  range: SourceRange;
};

export const parseError = (
  message: string,
  range: SourceRange,
): ParseErrorEntry => ({
  message,
  range,
});

export class ParseErrorStack {
  public static from(errors: ParseErrorEntry[]): ParseErrorStack {
    return new ParseErrorStack(errors);
  }

  private constructor(private errors: ParseErrorEntry[]) {}

  public all(): ParseErrorEntry[] {
    return this.errors;
  }

  public clear(): ParseErrorStack {
    const stack = ParseErrorStack.from([]);
    return stack;
  }

  public latest(): ParseErrorEntry | null {
    return this.errors[this.errors.length - 1] ?? null;
  }

  public upsert(
    from: ParseErrorEntry | null,
    to: ParseErrorEntry,
  ): ParseErrorStack {
    if (from == null) {
      return ParseErrorStack.from([...this.errors, to]);
    }

    const index = this.errors.findIndex((entry) => entry === from);
    if (index < 0) {
      return ParseErrorStack.from([...this.errors, to]);
    }

    return ParseErrorStack.from([
      ...this.errors.slice(0, index),
      to,
      ...this.errors.slice(index + 1),
    ]);
  }

  public append(entry: ParseErrorEntry): ParseErrorStack {
    const stack = ParseErrorStack.from([...this.errors, entry]);
    return stack;
  }

  public [Symbol.iterator]() {
    let index = -1;
    return {
      next: () => ({
        value: this.errors[++index],
        done: index === this.errors.length,
      }),
    };
  }
}
