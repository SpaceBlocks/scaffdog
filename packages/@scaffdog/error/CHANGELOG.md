# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.0.0-canary.6](https://github.com/cats-oss/scaffdog/compare/v2.0.0-canary.5...v2.0.0-canary.6) (2022-08-06)

**Note:** Version bump only for package @scaffdog/error

# [2.0.0-canary.5](https://github.com/cats-oss/scaffdog/compare/v2.0.0-canary.4...v2.0.0-canary.5) (2022-08-06)

**Note:** Version bump only for package @scaffdog/error

# [2.0.0-canary.4](https://github.com/cats-oss/scaffdog/compare/v2.0.0-canary.3...v2.0.0-canary.4) (2022-08-06)

**Note:** Version bump only for package @scaffdog/error

# [2.0.0-canary.3](https://github.com/cats-oss/scaffdog/compare/v2.0.0-canary.2...v2.0.0-canary.3) (2022-08-06)

**Note:** Version bump only for package @scaffdog/error

# [2.0.0-canary.2](https://github.com/cats-oss/scaffdog/compare/v2.0.0-canary.1...v2.0.0-canary.2) (2022-08-06)

**Note:** Version bump only for package @scaffdog/error

# [2.0.0-canary.1](https://github.com/cats-oss/scaffdog/compare/v2.0.0-canary.0...v2.0.0-canary.1) (2022-08-06)

**Note:** Version bump only for package @scaffdog/error

# [2.0.0-canary.0](https://github.com/cats-oss/scaffdog/compare/v1.5.0...v2.0.0-canary.0) (2022-08-06)

### Features

- **engine:** renew template engine ([4553a25](https://github.com/cats-oss/scaffdog/commit/4553a25bbae43c38dcaebb97cbe28fa4a24a16fa))

# [1.5.0](https://github.com/cats-oss/scaffdog/compare/v1.4.0...v1.5.0) (2022-06-26)

**Note:** Version bump only for package @scaffdog/error

# [1.4.0](https://github.com/cats-oss/scaffdog/compare/v1.3.0...v1.4.0) (2022-06-23)

**Note:** Version bump only for package @scaffdog/error

# [1.3.0](https://github.com/cats-oss/scaffdog/compare/v1.2.1-canary.1...v1.3.0) (2022-06-22)

**Note:** Version bump only for package @scaffdog/error

## [1.2.1-canary.0](https://github.com/cats-oss/scaffdog/compare/v1.2.0...v1.2.1-canary.0) (2022-06-21)

**Note:** Version bump only for package @scaffdog/error

# [1.2.0](https://github.com/cats-oss/scaffdog/compare/v1.1.0...v1.2.0) (2022-01-25)

**Note:** Version bump only for package @scaffdog/error

# [1.1.0](https://github.com/cats-oss/scaffdog/compare/v1.1.0-canary.0...v1.1.0) (2021-12-19)

**Note:** Version bump only for package @scaffdog/error

# [1.1.0-canary.0](https://github.com/cats-oss/scaffdog/compare/v1.0.1...v1.1.0-canary.0) (2021-12-19)

**Note:** Version bump only for package @scaffdog/error

## [1.0.1](https://github.com/cats-oss/scaffdog/compare/v1.0.1-canary.0...v1.0.1) (2021-04-11)

**Note:** Version bump only for package @scaffdog/error

## [1.0.1-canary.0](https://github.com/cats-oss/scaffdog/compare/v1.0.0...v1.0.1-canary.0) (2021-04-11)

### Bug Fixes

- move @scaffdog/types to dependencies from devDependencies ([8ee5eed](https://github.com/cats-oss/scaffdog/commit/8ee5eedd59bf8e0ccbf12a3884e662fe387980b6))

### Features

- **error:** add `format` method ([a161ffd](https://github.com/cats-oss/scaffdog/commit/a161ffd8f7620a12c3348a45350f641e7a8ae48d))

# [1.0.0](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.8...v1.0.0) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.8](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.7...v1.0.0-canary.8) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.7](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.6...v1.0.0-canary.7) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.6](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.5...v1.0.0-canary.6) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.5](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.4...v1.0.0-canary.5) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.4](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.3...v1.0.0-canary.4) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.3](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.2...v1.0.0-canary.3) (2021-01-12)

**Note:** Version bump only for package @scaffdog/error

# [1.0.0-canary.1](https://github.com/cats-oss/scaffdog/compare/v1.0.0-canary.0...v1.0.0-canary.1) (2021-01-12)

### Features

- improve error message ([4b66723](https://github.com/cats-oss/scaffdog/commit/4b667232ecd6cef7e33a145e7d0f35e8afb2fb81))

# [1.0.0-canary.0](https://github.com/cats-oss/scaffdog/compare/v0.3.0...v1.0.0-canary.0) (2021-01-11)

### Features

- v1 ([4080ec1](https://github.com/cats-oss/scaffdog/commit/4080ec14ef4397bd9061afca92eaf742926e58ac))
