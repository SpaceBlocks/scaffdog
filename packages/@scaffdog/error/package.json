{
  "name": "@scaffdog/error",
  "version": "1.0.0-canary.4",
  "description": "A utility module of scaffdog error.",
  "keywords": [
    "scaffdog",
    "markdown",
    "template"
  ],
  "homepage": "https://github.com/cats-oss/scaffdog/tree/main/packages/@scaffdog/error",
  "bugs": {
    "url": "https://github.com/cats-oss/scaffdog/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/cats-oss/scaffdog.git",
    "directory": "packages/@scaffdog/error"
  },
  "license": "MIT",
  "author": "wadackel <wadackel@gmail.com>",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!lib/**/*.test.js"
  ],
  "scripts": {
    "build": "tsc",
    "test": "ava"
  },
  "dependencies": {
    "chalk": "^4.1.0"
  },
  "devDependencies": {
    "@scaffdog/types": "1.0.0-canary.4"
  },
  "publishConfig": {
    "access": "public"
  }
}
