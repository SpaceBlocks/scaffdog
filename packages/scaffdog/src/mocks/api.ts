import { vi } from 'vitest';
import type { Scaffdog, ScaffdogInitializer } from '../api';
import { createResolvedConfig } from '../lib/config.factory';

export const createScaffdogMock = (
  mocks: Partial<Scaffdog> = {},
): Scaffdog => ({
  path: {
    project: '',
    config: '',
  },
  config: createResolvedConfig(),
  list: vi.fn(),
  generate: vi.fn(),
  ...mocks,
});

export const createScaffdogInitializerMock = (
  mocks: Partial<ScaffdogInitializer> = {},
): ScaffdogInitializer => {
  const scaffdog = createScaffdogMock();

  return {
    createScaffdog: vi.fn().mockReturnValueOnce(scaffdog),
    loadScaffdog: vi.fn().mockResolvedValueOnce(scaffdog),
    ...mocks,
  };
};
