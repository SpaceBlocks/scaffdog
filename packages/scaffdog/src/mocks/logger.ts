import { PassThrough } from 'stream';
import type { ConsolaInstance } from 'consola';
import { LogLevels, createConsola } from 'consola';
import stripAnsi from 'strip-ansi';

export const createLogger = (): {
  logger: ConsolaInstance;
  getStdout: () => string;
  getStderr: () => string;
} => {
  const output = {
    stdout: '',
    stderr: '',
  };

  const stream = {
    stdout: new PassThrough(),
    stderr: new PassThrough(),
  };

  const logger = createConsola({
    fancy: true,
    formatOptions: {
      date: false,
    },
    level: LogLevels.info,
    ...(stream as any),
  });

  stream.stdout.on('data', (chunk) => (output.stdout += chunk));
  stream.stderr.on('data', (chunk) => (output.stderr += chunk));

  return {
    logger,
    getStdout: () => stripAnsi(output.stdout),
    getStderr: () => stripAnsi(output.stderr),
  };
};
