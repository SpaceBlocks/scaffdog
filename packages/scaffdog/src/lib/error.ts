import { ScaffdogError } from '@scaffdog/error';
import type { Consola } from 'consola';
import indent from 'indent-string';
import { z } from 'zod';

export type ErrorLibrary = {
  handle: (e: unknown, title: string) => number;
};

export const createErrorLibrary = (logger: Consola): ErrorLibrary => ({
  handle: (e, title) => {
    if (e instanceof ScaffdogError) {
      logger.error(title);
      logger.log(indent(e.format({ color: true }), 4));
    } else if (e instanceof z.ZodError) {
      const msg = [`${e.errors.length} issues found`];
      for (const err of e.errors) {
        const code = err.code.toUpperCase();
        const path = err.path.join('.');
        msg.push(`- ${code}: ${err.message} (in "${path}")`);
      }
      logger.error(title);
      logger.log(indent(msg.join('\n'), 4));
    } else if (e instanceof Error) {
      e.message = `${title}: ${e.message}`;
      logger.error(e);
    } else {
      logger.error(title, e);
    }
    logger.log('');
    return 1;
  },
});

createErrorLibrary.inject = ['logger'] as const;
