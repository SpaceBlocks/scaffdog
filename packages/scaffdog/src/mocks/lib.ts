import type { Library } from '../lib';
import { createConfigLibraryMock } from '../lib/config.mock';
import { createDocumentLibraryMock } from '../lib/document.mock';
import { createErrorLibraryMock } from '../lib/error.mock';
import { createFsLibraryMock } from '../lib/fs.mock';
import { createPromptLibraryMock } from '../lib/prompt.mock';
import { createQuestionLibraryMock } from '../lib/question.mock';

export const createLibraryMock = (mocks: Partial<Library> = {}): Library => {
  return {
    fs: createFsLibraryMock(),
    error: createErrorLibraryMock(),
    prompt: createPromptLibraryMock(),
    config: createConfigLibraryMock(),
    question: createQuestionLibraryMock(),
    document: createDocumentLibraryMock(),
    ...mocks,
  };
};
