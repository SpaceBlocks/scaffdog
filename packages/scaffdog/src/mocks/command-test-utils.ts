import path from 'path';
import type {
  AnyCommandModule,
  CommandArgs,
  CommandModule,
  CommandOption,
  GlobalFlags,
  GlobalFlagsWith,
  InferredCommandModuleArgs,
  InferredCommandModuleFlags,
} from '../command';
import type { CommandContainer } from '../command-container';
import { createCommandContainer } from '../command-container';
import type { Library } from '../lib';
import { createLibraryMock } from './lib';
import { createLogger } from './logger';

export const cwd = path.resolve(__dirname, '../../');

export type CommandRunnerParams<T extends AnyCommandModule> = {
  args?: Partial<CommandArgs<InferredCommandModuleArgs<T>>>;
  flags?: Partial<GlobalFlagsWith<CommandArgs<InferredCommandModuleFlags<T>>>>;
  lib?: Library;
  container?: CommandContainer;
};

export type CommandRunner<T extends AnyCommandModule> = (
  params: CommandRunnerParams<T>,
) => Promise<{
  code: number;
  stdout: string;
  stderr: string;
}>;

export const createCommandRunner =
  <A extends CommandOption, F extends CommandOption>(
    cmd: CommandModule<A, F>,
    defaults: {
      args: CommandArgs<A>;
      flags: CommandArgs<F>;
    },
  ): CommandRunner<CommandModule<A, F>> =>
  async ({ args, flags, lib, container }) => {
    const { logger, getStdout, getStderr } = createLogger();

    const globalFlags: GlobalFlags = {
      project: '.scaffdog',
      help: false,
      version: false,
      verbose: true,
    };

    const code = await cmd.run({
      cwd,
      pkg: {
        name: 'scaffdog',
        version: '1.0.0',
        description: 'scaffdog is Markdown driven scaffolding tool.',
      },
      size: {
        rows: 120,
        columns: 60,
      },
      logger,
      container: container ?? createCommandContainer([]),
      lib: lib ?? createLibraryMock({}),
      args: {
        ...defaults.args,
        ...(args ?? {}),
      },
      flags: {
        ...defaults.flags,
        ...globalFlags,
        ...(flags ?? {}),
      } as never,
    });

    return {
      code,
      stdout: getStdout(),
      stderr: getStderr(),
    };
  };
