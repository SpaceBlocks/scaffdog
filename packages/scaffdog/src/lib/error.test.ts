import path from 'path';
import { ScaffdogError } from '@scaffdog/error';
import { describe, expect, test } from 'vitest';
import { z } from 'zod';
import { createLogger } from '../mocks/logger';
import { createErrorLibrary } from './error';

const ROOT_PATH = path.join(__dirname, '../../../..');

describe('handle', () => {
  const createLib = () => {
    const { logger, getStdout, getStderr } = createLogger();
    const lib = createErrorLibrary(logger);
    return {
      lib,
      logger,
      getStdout,
      getStderr,
    };
  };

  test.each([
    [
      'ScaffdogError - without source',
      new ScaffdogError('message', {}),
      'Compile Error',
    ],

    [
      'ScaffdogError - with source',
      new ScaffdogError('"foo" is not defined', {
        source: `{{ foo }}`.trim(),
        range: [3, 5],
      }),
      'Compile Error',
    ],

    [
      'ZodError',
      (
        z
          .object({
            foo: z.string(),
            bar: z.object({
              baz: z.number(),
            }),
          })
          .safeParse({
            foo: 0,
            bar: {
              baz: '',
            },
            qux: false,
          }) as any
      ).error,
      'Parsing Error',
    ],

    ['Error', new Error('title'), 'General Error'],

    ['unknown', null, 'Error'],
  ])('%s', (_, e, title) => {
    const { lib, getStdout, getStderr } = createLib();
    const code = lib.handle(e, title);
    const output = (getStderr() + getStdout()).replace(
      new RegExp(ROOT_PATH, 'g'),
      '~',
    );

    expect(code).toBe(1);
    expect(output).toMatchSnapshot();
  });
});
