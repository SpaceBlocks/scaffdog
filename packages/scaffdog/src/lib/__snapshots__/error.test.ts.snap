// Vitest Snapshot v1

exports[`handle > Error 1`] = `
"
 ERROR  General Error: title

  at src/lib/error.test.ts:55:15
  at ~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-chain.0ab05798.mjs:2319:43
  at runWithSuite (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-chain.0ab05798.mjs:2128:9)
  at Object.collect (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-chain.0ab05798.mjs:2319:13)
  at Object.collect (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-chain.0ab05798.mjs:2322:57)
  at async collectTests (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-error.f5506f40.mjs:351:28)
  at async startTestsNode (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-error.f5506f40.mjs:781:17)
  at async ~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/entry.mjs:83:9
  at async Module.withEnv (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/chunk-runtime-error.f5506f40.mjs:293:5)
  at async run (~/node_modules/.pnpm/vitest@0.23.4/node_modules/vitest/dist/entry.mjs:76:5)
  at async ~/node_modules/.pnpm/tinypool@0.3.0/node_modules/tinypool/dist/esm/worker.js:109:20


"
`;

exports[`handle > ScaffdogError - with source 1`] = `
"
 ERROR  Compile Error

       \\"foo\\" is not defined:

     > │ {{ foo }}
       │    ^^^

"
`;

exports[`handle > ScaffdogError - without source 1`] = `
"
 ERROR  Compile Error

    message

"
`;

exports[`handle > ZodError 1`] = `
"
 ERROR  Parsing Error

    2 issues found
    - INVALID_TYPE: Expected string, received number (in \\"foo\\")
    - INVALID_TYPE: Expected number, received string (in \\"bar.baz\\")

"
`;

exports[`handle > unknown 1`] = `
"
 ERROR  Error null


"
`;
